File: src/main/java/org/pl/parser/IParser.java
package org.pl.parser;

import org.pl.lexer.token.IToken;
import org.pl.parser.ast.INode;

import java.util.List;

public interface IParser {
    INode parse(List<IToken> tokens);

    default int advanceCursor() {
        return advanceCursor(1);
    }

    int advanceCursor(int by);

    int getCursorPosition();

    boolean hasNextToken();

    IToken nextToken();

    IToken getToken();
}

File: src/main/kotlin/org/leaflang/parser/syntax/IParser.kt
package org.leaflang.parser.syntax

import org.leaflang.parser.ast.INode


/**
 * An evaluator evaluates the next token(s) and returns a corresponding node.
 */
interface IParser {

    /**
     * Evaluates the next tokens.
     */
    fun parse(): INode
}